File: imooc_web/src/main/java/com/imooc/dao/CourseClickCountDAO.java
package com.imooc.dao;

import com.imooc.domain.CourseClickCount;
import com.imooc.utils.HBaseUtils;
import org.springframework.stereotype.Component;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/**
 * 实战课程访问数量数据访问层
 */
@Component
public class CourseClickCountDAO
{


    /**
     * 根据天查询
     */
    public List<CourseClickCount> query(String day) throws Exception
    {

        List<CourseClickCount> list = new ArrayList<>();

        // 去HBase表中根据day获取实战课程对应的访问量
        Map<String, Long> map = HBaseUtils.getInstance().query("boy_imooc_course_clickcount", "20180622");

        for (Map.Entry<String, Long> entry : map.entrySet())
        {
            CourseClickCount model = new CourseClickCount();
            model.setName(entry.getKey());
            model.setValue(entry.getValue());

            list.add(model);
        }

        return list;
    }

    public static void main(String[] args) throws Exception
    {
        CourseClickCountDAO dao = new CourseClickCountDAO();
        List<CourseClickCount> list = dao.query("20180622");
        for (CourseClickCount model : list)
        {
            System.out.println(model.getName() + " : " + model.getValue());
        }

        /**结果：
         * 20180622_128 : 351
         * 20180622_130 : 366
         * 20180622_131 : 332
         * 20180622_112 : 354
         * 20180622_145 : 348
         * 20180622_146 : 348
         */
    }

}

File: sparktrain/src/main/scala/com/imooc/spark/project/dao/CourseClickCountDAO.scala
package com.imooc.spark.project.dao

import com.imooc.spark.project.domain.CourseClickCount
import com.imooc.spark.project.utils.HBaseUtils
import org.apache.hadoop.hbase.client.Get
import org.apache.hadoop.hbase.util.Bytes

import scala.collection.mutable.ListBuffer


/**
  * 实战课程点击数-数据访问层
  */
object CourseClickCountDAO {

  val tableName = "boy_imooc_course_clickcount"
  //表名
  val cf = "info"
  //列簇
  val qualifer = "click_count" //列名

  /**
    * 保存数据到HBase
    *
    * @param list CourseClickCount集合
    */
  def save(list: ListBuffer[CourseClickCount]): Unit = {

    val table = HBaseUtils.getInstance().getTable(tableName)

    //incrementColumnValue:根据已有ID和新值ID进行累加
    for (ele <- list) {
      table.incrementColumnValue(Bytes.toBytes(ele.day_course),
        Bytes.toBytes(cf),
        Bytes.toBytes(qualifer),
        ele.click_count)
    }
  }


  /**
    * 根据rowkey查询值
    */
  def count(day_course: String): Long = {

    val table = HBaseUtils.getInstance().getTable(tableName)

    val get = new Get(day_course.getBytes())
    val value = table.get(get).getValue(cf.getBytes, qualifer.getBytes)

    if (value == null) {
      0L
    } else {
      Bytes.toLong(value)
    }
  }


  def main(args: Array[String]): Unit = {

    val list = new ListBuffer[CourseClickCount]
    list.append(CourseClickCount("20171111_8", 8))
    list.append(CourseClickCount("20171111_7", 81))
    list.append(CourseClickCount("20171111_9", 9))
    list.append(CourseClickCount("20171111_1", 34))

    save(list)

    println(count("20171111_8") + " : " + count("20171111_7") + " : " + count("20171111_9") + " : " + count("20171111_1"))

  }

}